lib/skywalking/tracer.lua

```diff
@@ -79,21 +79,26 @@ function Tracer:start(upstream_name, correlation)
     ctx.tracingContext = tracingContext
     ctx.entrySpan = entrySpan
     ctx.exitSpan = exitSpan
+    ctx.is_finished = false
 end
 
 function Tracer:finish()
     -- Finish the exit span when received the first response package from upstream
-    if ngx.ctx.exitSpan ~= nil then
+    if ngx.ctx.exitSpan ~= nil and not ngx.ctx.is_finished then
         local upstream_status = tonumber(ngx.var.upstream_status)
         if upstream_status then
             Span.tag(ngx.ctx.exitSpan, 'http.status', upstream_status)
         end
         Span.finish(ngx.ctx.exitSpan, ngx.now() * 1000)
```

Comment:
According to https://github.com/apache/apisix/issues/4301#issuecomment-847912259, we need to update the timestamp of the exit span if `#finish` method is called again.

Reply:
> the timestamp of the exit span

I see the timestamp of the exit span is real-time update when `#finish` method is called at L92.
Do you mean this?

Reply:
Yes, but you are going to call finish in the `header` first, then, the duration doesn't include the time of receiving body. Right? We need a way to update the end time as accurate as possible.

Reply:
Sorry for reply later.

With the way of add a flag to makesure `#finish` method will be called, it's no need to call finish in the `header_filter_by_lua` phase. Just use this project according to the origin doc.

I rearranged the logic. There is two way to fix this bug.
1. Add a flag to makesure `#finish` method must will be called. If `#finish` method not be called in `body_filter_by_lua` phase, it's will be called in `log_by_lua` phase. It's the way I take now.
2. Call `#finish` method in the `header_filter_by_lua` phase first. Then call `#finish` method in the `body_filter_by_lua` phase after that, updating the timestamp of the exit span at the same time. With this way there is no need to add flag, because `header_filter_by_lua` phase will always be execcute. But this will change a lot of the code.

The way 1 is the simplest way to fix this issue, and no serious problems will arise. So I take way 1.

Reply:
Make sense to me.